jep/0000/README.adoc

```diff
@@ -0,0 +1,309 @@
+- This behavior is transparent to plugin developers, not making it clear what goes on behind the scenes.
+
+While allowing selective access to files on the controller may have been a good solution in 2014 for compatibility with then-existing plugins, few plugins seem to need this exception today.
+
+Plugins should be restructured to not have agent-to-controller access where possible, or implement a `SlaveToMasterCallable` with explicit input validation as described in https://www.jenkins.io/doc/developer/security/remoting-callables/[the developer documentation] instead of relying on `FilePath`.
+
+
+=== Deprecation of `SlaveToMasterFileCallable`
+
+While `SlaveToMasterCallable` is needed for some use cases, `SlaveToMasterFileCallable` exists for convenience only (as an argument to `FilePath#act`), and relies on the nontrivial custom (de)serialization of `FilePath`.
+To discourage the creation of new (`File`)`Callable` in the agent-to-controller direction, and make it easier to reason about security of any (`File`)`Callable` sent through a remoting channel, this type is deprecated, and warnings are logged whenever it is deserialized on a controller.
+
+
+== Backwards Compatibility
+
+https://github.com/jenkins-infra/usage-in-plugins[`usage-in-plugins`] is used to check access to any of the types removed or substantially altered:
+
+----
+# General
+jenkins/security/s2m/AdminWhitelistRule
+jenkins/security/s2m/ConfigDirectory
+jenkins/security/s2m/ConfigFile
+jenkins/security/s2m/MasterKillSwitchConfiguration
+jenkins/security/s2m/MasterKillSwitchWarning
+# FilePathFilter
+jenkins/security/s2m/AdminFilePathFilter
+jenkins/security/s2m/DefaultFilePathFilter
+jenkins/security/s2m/FilePathRuleConfig
+jenkins/security/s2m/FilePathRule
+jenkins/security/s2m/OpMatcher
+jenkins/security/s2m/RunningBuildFilePathFilter
+jenkins/ReflectiveFilePathFilter
+jenkins/SoloFilePathFilter
+jenkins/ReflectiveFilePathFilter
+jenkins/FilePathFilterAggregator
+jenkins/FilePathFilter
+# Callables
+jenkins/security/s2m/AdminCallableMonitor
+jenkins/security/s2m/AdminCallableWhitelist
+jenkins/security/s2m/CallableDirectionChecker
+jenkins/security/s2m/CallableRejectionConfig
+jenkins/security/s2m/CallableWhitelist
+jenkins/security/s2m/CallableWhitelistConfig
+jenkins/security/s2m/RejectedCallable
+----
+
+The only plugin distributed by the Jenkins project that is using any of these types is https://plugins.jenkins.io/configuration-as-code/[Configuration as Code], which uses `AdminWhitelistRule` in https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/plugin/src/main/java/io/jenkins/plugins/casc/core/AdminWhitelistRuleConfigurator.java[`AdminWhitelistRuleConfigurator`].
```

Comment:
Do we have a PR to deprecate/delete this configurator?

Reply:
No, seems unnecessary to have that ready quickly as it will just emit log messages.

I expect that this is fairly commonly set in `init.groovy.d` use cases as well, so I'm not in a hurry to fully delete.